p2p: add initial shred.md (#23)
solana-foundation/specs: p2p/shred.md

# Shred packets

> The Solana Turbine protocol propagates confirmed data (blocks) within peer-to-peer networks in the form of shred packets.

Reviewer: ```suggestion
Shred packets are fragments of blocks (transaction data) that facilitate propagation of the blocks within peer-to-peer networks.
```
Feel free to tweak more as you see fit; some justification for the change
- I dropped confirmed as at the initial turbine broadcast I'd consider the status of the block to be "proposed"
- I dropped mention of turbine since we also transmit shreds around with repair (I think we consider repair to be distinct from turbine)

Data shreds are created by splitting up block data.
Code shreds are created by constructing erasure codes over the vector of data shreds.

All shreds are signed by the block producer.

## Revisions

Multiple revisions of the shred structures exist.
The shred revision is not explicitly encoded.

### v1: Genesis revision

Solana mainnet-beta genesis launched with support for code and data shreds with the legacy authentication mechanism.

### v2: Explicit data sizes

Breaking change to the data shred header adding a new size field at offset `0x56`.

The shred paylod shifts from offset `0x56` to `0x58`.

### v3: Merkle authentication mechanism

Introduction of two additional shred variants with the Merkle authentication scheme.

## Protocol

### Transport

> Shreds are transmitted via unidirectional streams.

Reviewer: Shred transmission is definitely unidirectional for initial turbine broadcast, but repairs involve a node explicitly asking to be sent specific shreds.

I'm wondering if we should drop the Transport section and save that for a separate spec on turbine? Ie this spec could cover the shreds (layout, data v coding, merkle, etc), and then turbine doc could reference this spec saying turbine acts on shreds?

If you agree with this sentiment, I would say to leave in the note in `Packet Size` section as that provides the justification for why that constant is that value.

> Each shred is transmitted in the payload of an UDP packet.

Reviewer 2: ```suggestion
Each shred is transmitted in the payload of a UDP packet.
```

Otherwise, set to one to mark the end of the current batch.

### Data Shred Header (rev. v1)

Revision v1 data shreds.

The total size field is hardcoded to 1228 bytes.

| Offset | Size | Type  | Name           |  Purpose                       |
|--------|-----:|-------|----------------|--------------------------------|
| `0x53` |   4B | `u32` | `parent_dist`  |  Slot distance to parent block |
| `0x55` |   1B | `u8`  | `data_flags`   |  Data Flags                    |
| `0x56` |    - | bytes | `data_payload` |  Data Payload                  |

## Code Shreds

## Legacy Authentication

In the legacy authentication scheme, the block producer signs the content of each shred.

The message to be signed begins 64 bytes past the beginning of the common shred header (skipping the signature field) and includes the entire rest of the shred.
The resulting signature is placed into the _block producer signature_ field.

## Merkle Authentication

In the Merkle authentication scheme, the block producer signs the Merkle root over the FEC set of shreds.
Consequently, the signature field carries identical content for all shreds in the same FEC set.

> The Merkle node size is truncated to 20 bytes.

Reviewer 2: Merkle hashes only 20bytes?

## Shred Construction

Block producers create and broadcast shreds to the validator network while active.

This occurs as a streaming process wherein shreds are constructed in the earliest possible opportunity.

The construction of shreds requires the following steps:
1. Block Entry Batching
2. Data Shredding
3. Erasure Coding
4. Signing

## Block Entry Batching

> Batching creates sub-groups of block entries, each of which are serialized into one byte array.

Reviewer 2: it's worth noting that entries are non-contentious batches of txs, this way clients are inclined to make this assumption

  \begin{cases}
    F_i = \text{make\_data\_shreds}\left(
        B, 0,
        \text{serialize\_entry\_batch}(E_i) \right)
      &\text{if }i = 0\\
    F_i = \text{make\_data\_shreds}\left(
        B, I_{i-1} + 1,
        \text{serialize\_entry\_batch}(E_i) \right)
      &\text{if }i > 0\\
  \end{cases} \\
  ∀F_i ∈ F :
  I_i = \text{property}\left( \text{last}({F_i}),
                              \text{\textquotedblleft{shred\_index}\textquotedblright} \right)
  $
  <br/> Each data shred batch is the result of splitting up the corresponding serialized entry shred batch, given various block info, and the shred start index.
  <br/> The shred start index is zero for the first batch, and one plus the last shred index of the preceding batch for subsequent batches.

- $∀b ∈ F : ∀x ∈ [0;\lvert b \rvert) : ∀y ∈ [0;\lvert b \rvert) : y-x =
  \text{property}\left( b_y,
                        \text{\textquotedblleft{shred\_index}\textquotedblright} \right)-
  \text{property}\left( b_x,
                        \text{\textquotedblleft{shred\_index}\textquotedblright} \right)$
  <br/> Shred indices within each shred batch are incrementing

- $S = \mathbin\Vert_{i=0}^{\lvert F \rvert} F_i$

>   <br/> Vector of shreds is the concatenation of all shred batch vectors

Reviewer 2: some of the formatting messed up here

- $∀i ∈ [0;\lvert S \rvert) : i = \text{property} \left( {S_i}, \text{\textquotedblleft{shred\_index}\textquotedblright} \right)$
  <br/> Shred indices are incrementing, starting at zero

- $∀F_i ∈ F : F_i ⊆ S$
  <br/> Each data shred batch is a subset of the data shred vector

- $
  ∀F_i ∈ F : ∀F_j ∈ F :
  \begin{cases}
    F_i = F_j             & \text{if } i = j \\
    F_i ⋂ F_j = \emptyset & \text{if } i \neq j
  \end{cases}
  $

>   <br/> Data shred batches are non-overlapping

Reviewer 2: ah I see it. it this true? the way I understand it is:
```
non-overlapping tx batch  -> entry -> shred
```
